### PR TITLE
lock RefreshToken entity with pissimistic locking to avoid concurrent updates

### PR DESCRIPTION
**Pull request by vgagara-talend**
issue: https://issues.apache.org/jira/browse/CXF-7374
orig issue: https://jira.talendforge.org/browse/TPSVC-2439

### CODE REVIEW

**Comment by sberyozkin** [comment]
Hi, thanks for the patch. Can you please consider locking at the AbstractOAuthDataProvider level instead, so that all its extensions (JPA2, JCache, Ehcache and future ones like Mongo-based) can benefit ?
I'd consider introducing 'Object refreshTokenLock' and use it lock at https://github.com/apache/cxf/blob/master/rt/rs/security/oauth-parent/oauth2/src/main/java/org/apache/cxf/rs/security/oauth2/provider/AbstractOAuthDataProvider.java#L211

Let me know please if it works for you

File: rt/rs/security/oauth-parent/oauth2/src/main/java/org/apache/cxf/rs/security/oauth2/provider/JPAOAuthDataProvider.java [hunk 1]
```diff
@@ -213,6 +221,25 @@ public RefreshToken execute(EntityManager em) {
...
 
+    protected void lockRefreshTokenForUpdate(final RefreshToken refreshToken) {
+        try {
+            execute(new EntityManagerOperation<Void>() {
+
+                @Override
+                public Void execute(EntityManager em) {
+                    Map<String, Object> options = Collections.emptyMap();
+                    if (pessimisticLockTimeout > 0) {
+                        Collections.singletonMap("javax.persistence.lock.timeout", pessimisticLockTimeout);
```
Comment (andrei-ivanov) [inline rt/rs/security/oauth-parent/oauth2/src/main/java/org/apache/cxf/rs/security/oauth2/provider/JPAOAuthDataProvider.java]: options = ...

Comment (sberyozkin) [inline rt/rs/security/oauth-parent/oauth2/src/main/java/org/apache/cxf/rs/security/oauth2/provider/JPAOAuthDataProvider.java]: Good catch, I'm applying PR now so will fix

**Comment by sberyozkin** [comment]
Viacheslav, the patch has been applied with minor updates to let people use a Java lock if they prefer (JPA2 provider is currently using the pessimistic lock as per your code).

However the test fails so I disabled it - can you please have a look and also copy it it to JPACMTOAuthDataProviderOpenJPATest class to make sure it works with OpenJPA - to make sure we don't break the JPA2 provider depending on OpenJPA ? Thanks

**Pull request closed without merging**